grails-app/services/org/pih/warehouse/report/NotificationService.groovy

```diff
@@ -118,6 +122,39 @@ class NotificationService {
 
     }
 
+    @EventListener
+    void shipmentNotification(ShipmentStatusTransitionEvent event){
```

Comment:
You can leave this now, but add a TODO somewhere to remind us to rename this class EmailNotificationService and move it out of the report package.

Reply:
I think we probably want an EventNotificationService that defines all of the onEvent handlers for anything we want to notify on (ShipmentStatusTransitionEvent, PostInsertEvent). We use this service to find out the event type ("shipment.shipped", "product.created") and use `openboxes.notification` config to determine what type of event publishing needs to happen. 

For example, let's say we have a `shipment.shipped` event. We look up in the event notification config to find anything that supports "shipment.shipped". We could get back config for SMTP and webhook. So your getWebhookNotificationConfig should just be getNotificationConfig in this EventNotificationService.

```
interface NotificationEvent {
    getEventType()
    getEventContext()
}

PostInsertNotificationEvent implements NotificationEvent { 
    PostInsertEvent event 
    getEventType() { return entity + ".created" }
    getEventContext() { 
        // TBD - not sure if this should be event.state or something else 
    }
}

class ShipmentStatusTransitionEvent implements NotificationEvent {

    Shipment shipment 
    ShipmentStatusCode shipmentStatusCode
    getEventType() { return shipmentStatusCode.eventType }
    getEventContext() { 
            return [
                  id            : shipment.id,
                  shipmentNumber: shipment.shipmentNumber,
                  origin        : shipment.origin?.id,
                  destination   : shipment.destination?.id,
                  shipmentType  : shipment.shipmentType,
                  shipmentItems : shipment.shipmentItems?.collect {
                      [
                              id            : it.id,
                              productName   : it.inventoryItem.product?.name,
                              productCode   : it.inventoryItem?.product.productCode,
                              lotNumber     : it?.inventoryItem?.lotNumber,
                              expirationDate: it?.inventoryItem?.expirationDate.format("MM/dd/yyyy"),
                              quantity      : it.quantity
                      ]
                  }
            ]
    }
```    
The system will still publish ShipmentStatusTransitionEvent in the same way as it did before, but now when we get any of these events we can handle them in a generic way. 

```
class NotificationService {
    onEvent(NotificationEvent notificationEvent)
        // return all enabled notification configurations
        getNotificationConfig(notificationEvent.eventType).each { Map notificationConfig ->
            switch (notificationConfig.provider) {
                case NotificationProvider.WEBHOOK: 
                    publishEvent(new WebhookEvent(notificationEvent.context, notificationConfig))
                case NotificationProvider.EMAIL: 
                    publishEvent(new SmtpEvent(notificationEvent.context, notificationConfig))
                default: 
                    log.warn("No available provider ${notificationConfig.provider} for event type ${notificationEvent.eventType")
            }
        }
    }
}
```
and the webhook publisher service would be way more simple
```
class WebhookNotificationService {
    def onWebhookEvent(WebhookEvent webhookEvent) { 
        String webhookUrl = webhookEvent.config.url
        Map headers = webhookEvent.config.headers
        log.info "Payload:$payload, webhookUrl:${webhookUrl}, headers:${headers}"
        apiClientService.post(webhookUrl, payload, headers)
    } 
}

class EmailNotificationService { 
    def onEmailEvent(EmailEvent emailEvent) {         
        mailService(emailEvent.subject, emailEvent.body, emailEvent.to)
    }
}

}
```

Reply:
What discuss on this half of things are implemented like reading yml file to fire webhooks and other event. created NotificationEventService